### DISCUSSION TITLE
Lesson 15 weak RNG question

### DISCUSSION THREAD

**Discussion by NeoRusi** [Cyfrin/foundry-full-course-cu#1071] [Q&A]
Hi guys i now this question may not be very related to the course, but its about foundry fuzz tests, and I couldn't find info about it ib the docs so writing here
<img width="825" alt="Screenshot 2023-10-31 at 18 19 35" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/92cf05c9-5c75-4502-b668-dd903010182f">
In the 'selectWinner' function I see a vulnerability weak RNG

in order to prove it I have written 2 smart contracts which simulate the work of this function( a bit )

here is the 1 a have changed ```block.difficulty``` and ```block.timestamp``` as uints256( maybe its incorrect) but I don't now any other way to fuzz their values to show that it is vulnreble 
```solidity 
//SPDX-License-Identifier: MIT

pragma solidity ^0.7.6;
pragma experimental ABIEncoderV2;

import {Test} from "forge-std/Test.sol";

contract HandleSelectWinner is Test {
    function selectWinnerRarity(uint256 block_difficulty) public returns (uint256 rarity) {
        address msg_sender = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
        rarity = uint256(keccak256(abi.encodePacked(msg.sender, block_difficulty))) % 100;
        return rarity;
    }

    function selectWinnerIndex(uint256 block_difficulty, uint256 block_timestamp)
        public
        returns (uint256 winnerIndex)
    {
        address msg_sender = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
        winnerIndex = uint256(keccak256(abi.encodePacked(msg_sender, block_timestamp, block_difficulty))) % 4;
        return winnerIndex;
    }
}
```
and here is the second
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.7.6;
pragma experimental ABIEncoderV2;

import {Test, console} from "forge-std/Test.sol";
import {StdInvariant} from "forge-std/StdInvariant.sol";
import {PuppyRaffle} from "../../src/PuppyRaffle.sol";
import {HandleSelectWinner} from "./HandleSelectWinner.t.sol";

contract PuppyRaffleTest is StdInvariant, Test {
    PuppyRaffle puppyRaffle;
    HandleSelectWinner handleSelectWinner;
    uint256 entranceFee = 1e18;
    uint256 block_difficulty;
    uint256 block_timestamp;
    address playerOne = address(1);
    address playerTwo = address(2);
    address playerThree = address(3);
    address playerFour = address(4);
    address feeAddress = address(99);
    uint256 duration = 1 days;

    function setUp() public {
        puppyRaffle = new PuppyRaffle(
            entranceFee,
            feeAddress,
            duration
        );
        handleSelectWinner = new HandleSelectWinner();
        targetContract(address(handleSelectWinner));
    }

    function invariant_testSelectWinnerIsNotRandom() public {
        // msg.sender --const
        // block.timestamp - is set up
        // block.difficulty is setting up
        console.log(handleSelectWinner.selectWinnerIndex(block_difficulty, block_timestamp));
        assert(handleSelectWinner.selectWinnerIndex(block_difficulty, block_timestamp) == 1);
    }

    function invariant_testSelectWinnerRarityIsNotRandom() public {
        // msg.sender -- anvil address
        // block.difficulty is variable
        // uint256(keccak256(abi.encodePacked(msg.sender, block.difficulty))) % 100 = 4
        // abi.decode ?
        console.log(handleSelectWinner.selectWinnerRarity(block_difficulty));
        assert(handleSelectWinner.selectWinnerRarity(block_difficulty) == 4);
    }
}
```

also for some reason I get this error about invalid opcodes !?
```
Failing tests:
Encountered 2 failing tests in test/fuzz/FuzzTests.t.sol:PuppyRaffleTest
[FAIL. Reason: EvmError: InvalidFEOpcode] invariant_testSelectWinnerIsNotRandom() (runs: 257, calls: 3840, reverts: 0)
[FAIL. Reason: EvmError: InvalidFEOpcode] invariant_testSelectWinnerRarityIsNotRandom() (runs: 257, calls: 3840, reverts: 0)
```

Its actually I quite funny situation I see the vulnerability but I can't prove it because have started not long ago 😂😂
Asked the question here cause it's related to the fuzzing course section :).

**Comment by Niferu** [comment]
Hello @NeoRusi 

You are totally right, provided code has vulnerability, you can even find challenge associated to it :) This have been made on purpose. To use truly random winner picker you should be using VRFConsumerBaseV2 contract.

**Reply by NeoRusi** [reply]
@Neftyr I know that you should use Chainlink VRF  to avoid this vulnerability, and you are right that there is a challenge, the question  I have is that I found 2 critical vulnerabilities which can be manipulated by miners, is it enough to mention this fact and put a link to the initial repo in submitting , or should I be proving it with a test, because if I should prove it with a test I have problems cause I maybe mistake in some parts of my code 🫠

**Reply by NeoRusi** [reply]
And what is good to mention, its very fun that as a developer I previously developed real dApps with money in them, and now when I try myself as an auditor its hard to write a single test for such an easy contract, and ChatGPT doesn't works with foundry 😥, but if I don't try I won't have results, so the only way is to continue 🚀

**Reply by NeoRusi** [reply]
Also, slither showed me a reentrancy in withdrawFees, but because it withdrawsFees to a address set internally(feeAddress) I believe its is okay ?

**Reply by Niferu** [reply]
Hi @NeoRusi 

In which exact contract you see that `selectWinner()` function? Regarding Slither he is yelling at everything, even if you put reentrancy guard to function, so I would not trust it fully.
